### PR TITLE
Fix CAST in TRY caused by peeling of input encoding

### PR DESCRIPTION
**Pull request by kagamiori**
Summary:
Casting from/to custom types and casting to complex types both peel off input encodings before processing the data. However, they only apply the encoding to the result vector but not error vector after the processing. This caused a mismatch of rows in the error vector and the result vector, and hence the incorrect result of `try(cast(...))` expressions.

This diff fix this bug by using ContextSaver. ContextSaver automatically apply peeled encodings back to the error vector during destruction. This diff also does a minor refactoring in Expr.cpp to reduce code duplicates.

Differential Revision: D39158842

### CODE REVIEW

File: velox/functions/prestosql/tests/JsonCastTest.cpp [hunk 1]
```diff
@@ -1000,3 +1000,11 @@ TEST_F(JsonCastTest, toInvalid) {
       {R"(["red",1.1])"_sv, R"(["blue",2.2])"_sv});
   testThrow<Json, ComplexType>(JSON(), ARRAY(ROW({DOUBLE()})), {"null"_sv});
 }
+
+TEST_F(JsonCastTest, castInTry) {
+  auto input = makeRowVector(
+      {makeNullableFlatVector<Json>({"1a"_sv, "2"_sv, "3"_sv}, JSON())});
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/JsonCastTest.cpp:1006]: nit: we can use makeFlatVector here (it would be shorter)

File: velox/functions/prestosql/tests/CastBaseTest.h [hunk 2]
```diff
@@ -35,19 +35,23 @@ class CastBaseTest : public FunctionBaseTest {
         std::make_unique<::facebook::velox::test::TestingDictionaryFunction>());
   }
 
+  core::TypedExprPtr buildCastExpr(
+      const TypePtr& fromType,
+      const TypePtr& toType,
+      bool tryCast = false) {
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/CastBaseTest.h:41]: Looks like this parameter is always passed, hence, no need to provide default value.

File: velox/functions/prestosql/tests/CastBaseTest.h [hunk 3]
```diff
@@ -67,25 +71,37 @@ class CastBaseTest : public FunctionBaseTest {
     assertEqualVectors(expected, result);
   }
 
+  core::TypedExprPtr buildCastExprDictEncoding(
+      const TypePtr& fromType,
+      const TypePtr& toType,
+      bool tryCast = false) {
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/CastBaseTest.h:77]: Same as above. No need for the default value.

File: velox/functions/prestosql/tests/CastBaseTest.h [hunk 4]
```diff
@@ -67,25 +71,37 @@ class CastBaseTest : public FunctionBaseTest {
...
+        "testing_dictionary");
+    return std::make_shared<const core::CastTypedExpr>(
+        toType, std::vector<core::TypedExprPtr>{callExpr}, tryCast);
+  }
+
+  void verifyCastDictEncoding(
+      const VectorPtr& expected,
+      const VectorPtr& actual) {
+    auto indices =
+        ::facebook::velox::test::makeIndicesInReverse(expected->size(), pool());
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/CastBaseTest.h:92]: Can we drop '::facebook::velox::test::' prefix?

File: velox/functions/prestosql/tests/CastBaseTest.h [hunk 5]
```diff
@@ -67,25 +71,37 @@ class CastBaseTest : public FunctionBaseTest {
...
+        std::make_shared<const core::FieldAccessTypedExpr>(fromType, "c0");
+    core::TypedExprPtr callExpr = std::make_shared<const core::CallTypedExpr>(
+        fromType,
+        std::vector<core::TypedExprPtr>{inputField},
+        "testing_dictionary");
+    return std::make_shared<const core::CastTypedExpr>(
+        toType, std::vector<core::TypedExprPtr>{callExpr}, tryCast);
+  }
+
+  void verifyCastDictEncoding(
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/CastBaseTest.h:88]: This method is a bit cryptic. It is confusing to pass 'expected' values, but then modify it by applying a very particular type of dictionary encoding. It might be clearer to just move this logic to the caller.

File: velox/functions/prestosql/tests/CastBaseTest.h [hunk 6]
```diff
@@ -67,25 +71,37 @@ class CastBaseTest : public FunctionBaseTest {
...
+      const VectorPtr& expected,
+      const VectorPtr& actual) {
+    auto indices =
+        ::facebook::velox::test::makeIndicesInReverse(expected->size(), pool());
+    assertEqualVectors(
+        wrapInDictionary(indices, expected->size(), expected), actual);
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/CastBaseTest.h:94]: Consider adding helper function wrapInDictionary that takes only indices and base vector and uses indices->size() / sizeof(vector_size_t) for size.

File: velox/functions/prestosql/tests/CastBaseTest.h [hunk 7]
```diff
@@ -94,10 +110,39 @@ class CastBaseTest : public FunctionBaseTest {
...
+  void evaluateCastInTry(
+      const TypePtr& fromType,
+      const TypePtr& toType,
+      const RowVectorPtr& input,
+      const VectorPtr& expected) {
+    auto castExpr = buildCastExpr(fromType, toType, false);
+    std::shared_ptr<const core::ITypedExpr> tryExpr =
+        std::make_shared<const core::CallTypedExpr>(
+            toType,
+            std::vector<std::shared_ptr<const core::ITypedExpr>>{castExpr},
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/CastBaseTest.h:125]: core:TypedExprPtr

File: velox/functions/prestosql/tests/CastBaseTest.h [hunk 8]
```diff
@@ -94,10 +110,39 @@ class CastBaseTest : public FunctionBaseTest {
...
+  void evaluateCastInTryDictEncoding(
+      const TypePtr& fromType,
+      const TypePtr& toType,
+      const RowVectorPtr& input,
+      const VectorPtr& expected) {
+    auto castExpr = buildCastExprDictEncoding(fromType, toType, false);
+    std::shared_ptr<const core::ITypedExpr> tryExpr =
+        std::make_shared<const core::CallTypedExpr>(
+            toType,
+            std::vector<std::shared_ptr<const core::ITypedExpr>>{castExpr},
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/CastBaseTest.h:141]: core::TypedExprPtr

File: velox/expression/tests/CastExprTest.cpp [hunk 9]
```diff
@@ -759,3 +728,14 @@ TEST_F(CastExprTest, decimalToDecimal) {
           DECIMAL(12, 0)),
       true);
 }
+
+TEST_F(CastExprTest, castInTry) {
+  auto data = makeVectorWithNullArrays<StringView>(
```
Comment (mbasmanova) [inline velox/expression/tests/CastExprTest.cpp:733]: Having both 'data' and 'input' makes this code a bit hard to read. Consider inlining 'data' variable.

**Review by mbasmanova — COMMENTED** [review]
@kagamiori Nice tests. Some comments below. I'll read the code changes next.

File: velox/expression/CastExpr.cpp [hunk 10]
```diff
@@ -520,63 +520,80 @@ VectorPtr CastExpr::applyDecimal(
...
-    DecodedVector& inputDecoded,
-    const SelectivityVector& allRows,
-    const SelectivityVector& nonNullRows,
-    const CastOperatorPtr& castOperator,
-    const TypePtr& thisType,
-    const TypePtr& otherType,
+/// Apply casting from a non-custom type to a complex type.
+/// @param decoded The decoded input vector
+/// @param rows The selectivity vector of non-null rows in decoded.base().
+void CastExpr::applyComplexType(
```
Comment (mbasmanova) [inline velox/expression/CastExpr.cpp:526]: Usually, when we pass 'rows' and 'decodedVector', 'rows' are top-level rows that can be used with decodedVector.isNullAt() and valueAt. Here, 'rows' are row numbers of the base vector. I wonder if it would be clearer to pass the base vector itself to this method instead of DecodedVector. I assume the logic here only ever uses decoded.base() and nothing else.

File: velox/expression/CastExpr.cpp [hunk 11]
```diff
@@ -520,63 +520,80 @@ VectorPtr CastExpr::applyDecimal(
...
+    const TypePtr& fromType,
+    const TypePtr& toType,
+    VectorPtr& result) {
+  if (castFromOperator_ || castToOperator_) {
+    VELOX_CHECK_NE(
+        fromType,
+        toType,
+        "Attempting to cast from {} to itself.",
+        fromType->toString());
+    context.ensureWritable(rows, toType, result);
```
Comment (mbasmanova) [inline velox/expression/CastExpr.cpp:585]: It would be nice to let the castToOperator decide whether they want to make a new 'result' vector or reuse the one provided. In other words, I think we should remove this call here and let castTo handle the 'result' similar to how VectorFunction::apply does.

File: velox/expression/CastExpr.cpp [hunk 12]
```diff
@@ -602,99 +619,59 @@ void CastExpr::apply(
...
           *decoded,
-          result);
+          context,
+          fromType,
+          toType,
+          localResult);
+
+      localResult = decoded->wrap(localResult, *input, rows);
     }
+    context.moveOrCopyResult(localResult, rows, result);
```
Comment (mbasmanova) [inline velox/expression/CastExpr.cpp:660]: We may need to call moveOrCopyResult after saver is destroyed. See https://github.com/facebookincubator/velox/pull/2389

CC: @bikramSingh91

**Comment by bikramSingh91** [comment]
Thanks @kagamiori for the change. I have unfortunately been preoccupied with making changes to the velox website today, but should be able to take out time tomorrow to review.

File: velox/expression/CastExpr.cpp (hunk 12 above)
Comment (kagamiori) [reply velox/expression/CastExpr.cpp:660]: I think the current code does call moveOrCopyResult after saver is destroyed. Not sure if I understand the concern. Could you clarify a bit?

Comment (mbasmanova) [reply velox/expression/CastExpr.cpp:658]: > I think the current code does call moveOrCopyResult after saver is destroyed.

I didn't notice this. Great. Thank you for pointing out.

File: velox/functions/prestosql/tests/CastBaseTest.h [hunk 13]
```diff
@@ -41,13 +49,7 @@ class CastBaseTest : public FunctionBaseTest {
       const TypePtr& toType,
       const RowVectorPtr& input,
       bool tryCast = false) {
```
Comment (bikramSingh91) [inline velox/functions/prestosql/tests/CastBaseTest.h:51]: nit: how about isTryCast

File: velox/functions/prestosql/tests/CastBaseTest.h [hunk 14]
```diff
@@ -67,25 +69,28 @@ class CastBaseTest : public FunctionBaseTest {
...
+  core::TypedExprPtr buildCastExprDictEncoding(
+      const TypePtr& fromType,
+      const TypePtr& toType,
+      bool tryCast) {
+    core::TypedExprPtr inputField =
+        std::make_shared<const core::FieldAccessTypedExpr>(fromType, "c0");
+    core::TypedExprPtr callExpr = std::make_shared<const core::CallTypedExpr>(
+        fromType,
+        std::vector<core::TypedExprPtr>{inputField},
+        "testing_dictionary");
```
Comment (bikramSingh91) [inline velox/functions/prestosql/tests/CastBaseTest.h:81]: nit: maybe rename to : "reverse_input_by_applying_dictionary_wrap"

File: velox/functions/prestosql/tests/CastBaseTest.h [hunk 15]
```diff
@@ -67,25 +69,28 @@ class CastBaseTest : public FunctionBaseTest {
     assertEqualVectors(expected, result);
   }
 
+  core::TypedExprPtr buildCastExprDictEncoding(
```
Comment (bikramSingh91) [inline velox/functions/prestosql/tests/CastBaseTest.h:72]: nit: how about buildCastExprWithDictionaryInput

**Review by bikramSingh91 — COMMENTED** [review]
@kagamiori Looks good, just a few nits

File: velox/expression/Expr.cpp [hunk 16]
```diff
@@ -509,6 +509,29 @@ void setDictionaryWrapping(
 }
 } // namespace
 
+void setConstantWrappingAndTranslateRows(
```
[the remark this answers is not in the document]
Comment (kagamiori) [reply velox/expression/Expr.cpp:512]: Yes, I agree and your proposal sounds good to me. I'll rename the added functions in this diff, and we can rename existing functions in a separate diff. Does this sound good?

File: velox/functions/prestosql/tests/CastBaseTest.h (hunk 14 above)
Comment (kagamiori) [reply velox/functions/prestosql/tests/CastBaseTest.h:81]: It's a bit long. What about "apply_reverse_dictionary"?

Comment (kagamiori) [reply velox/functions/prestosql/tests/CastBaseTest.h:81]: Actually, could we rename this function in a separate diff? This function is used not only in my tests here.

File: velox/expression/Expr.cpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (mbasmanova) [reply velox/expression/Expr.cpp:512]: I find the names and meaning of these function difficult to understand. These functions are quite small. I wonder if it make sense to just write out that code explicitly in the caller(s)?

> setDictionaryWrapping or setConstantWrapping sound like they are applying wrapping, whereas they are storing the state of a successful peeled operation. How about we use the term 'peel' instead of wrap to signify that we are storing a peeled wrap which then needs to be applied back.
eg setConstantPeelAndPeelRows , setDictionaryPeel
similarly also rename the functions and member variables in EvalCtx.

The naming can be improved indeed. Consider sending a PR to propose better names.

File: velox/expression/CastExpr.cpp [hunk 17]
```diff
@@ -520,63 +520,73 @@ VectorPtr CastExpr::applyDecimal(
...
+    } else {
+      castFromOperator_->castFrom(
+          input, context, rows, nullOnFailure_, toType, result);
+    }
   } else {
-    context.ensureWritable(*baseRows, otherType, localResult);
-
-    castOperator->castFrom(
-        *inputDecoded.base(), context, *baseRows, nullOnFailure, *localResult);
+    applyComplexType(rows, input, context, fromType, toType, result);
```
Comment (mbasmanova) [inline velox/expression/CastExpr.cpp:588]: applyComplexType method seems to be used only once. Consider folding it into the only caller.

File: velox/expression/CastExpr.cpp [hunk 18]
```diff
@@ -602,99 +612,64 @@ void CastExpr::apply(
...
-        case TypeKind::SHORT_DECIMAL:
-        case TypeKind::LONG_DECIMAL:
-          localResult = applyDecimal(
-              *translatedRows, *decoded, context, fromType, toType);
-          break;
-        default: {
-          VELOX_UNREACHABLE();
-        }
-      }
+  if (castFromOperator_ || castToOperator_ || !toType->isPrimitiveType() ||
```
Comment (mbasmanova) [inline velox/expression/CastExpr.cpp:615]: Why do we have different logic for different types? Shouldn't we use ContextSaver and the logic to translating errors on decoded rows to errors on original rows for all types?

At a high level, I believe the logic here is the same as in generic expression evaluation, i.e.

1. Identify null rows and set them aside.
2. Decode non-null inputs.
3. Evaluate cast logic on non-null flat inputs.
4. Set errors and wrap results based on (2). 
5. Add nulls based on (1).

It would be nice to structure the code to match this flow. it would also be nice to unify the code paths that process flat inputs between the cases where original inputs are flat and not.

File: velox/expression/CastExpr.cpp [hunk 19]
```diff
@@ -602,99 +612,64 @@ void CastExpr::apply(
...
+          *translatedRowsHolder,
+          *decoded->base(),
           context,
-          *decoded,
-          result);
+          fromType,
+          toType,
+          localResult);
+
+      localResult = decoded->wrap(localResult, *input, rows);
```
Comment (mbasmanova) [inline velox/expression/CastExpr.cpp:656]: Would it make sense to use EvalCtx::setWrapped here?

File: velox/vector/tests/VectorMaker.h [hunk 20]
```diff
@@ -494,7 +494,7 @@ class VectorMaker {
     auto rawOffsets = offsets->asMutable<vector_size_t>();
     auto rawSizes = sizes->asMutable<vector_size_t>();
     auto rawNulls = nulls->asMutable<uint64_t>();
-    bits::fillBits(rawNulls, 0, size, pool_);
```
Comment (mbasmanova) [inline velox/vector/tests/VectorMaker.h:497]: Nice catch. Would you extract this fix into a separate PR?

File: velox/expression/Expr.cpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (kagamiori) [reply velox/expression/Expr.cpp:512]: I created these functions because these logics are used in both Expr.cpp and CastExpr.cpp. I feel extracting and reusing them would make it easier to maintain the code in the future.

File: velox/expression/CastExpr.cpp (hunk 18 above)
Comment (kagamiori) [reply velox/expression/CastExpr.cpp:615]: This sounds nice, but it requires larger refactoring of the existing code because the current code of casting primitive types (except Decimal) does not peel off encodings before processing. Would it make sense to do this refactoring separately after this fix, or would you like it to be done before this fix?

File: velox/expression/CastExpr.cpp [hunk 21]
```diff
@@ -520,63 +520,73 @@ VectorPtr CastExpr::applyDecimal(
...
+      break;
+    case TypeKind::ROW:
+      result = applyRow(
+          rows,
+          input.asUnchecked<RowVector>(),
+          context,
+          fromType->asRow(),
+          toType->asRow());
+      break;
+    case TypeKind::SHORT_DECIMAL:
```
[the remark this answers is not in the document]
Comment (kagamiori) [reply velox/expression/CastExpr.cpp:556]: I agree. Decimal needs to take the same path with complex types and custom types (as opposed to other primitive types) because they all peel off encodings before processing the data, and thus I combined these paths to reduce code duplicates. What would you suggest as the name of a function that handles Decimal, complex, and custom types? (If we don't want to put them into one function, there will be some code duplicates.)

File: velox/expression/CastExpr.cpp (hunk 18 above)
Comment (mbasmanova) [reply velox/expression/CastExpr.cpp:615]: It would be nice to refactor before introducing material changes.

File: velox/expression/CastExpr.cpp (hunk 21 above)
[the remark this answers is not in the document]
Comment (mbasmanova) [reply velox/expression/CastExpr.cpp:556]: I wonder why primitive types (other than decimal) are handled differently. I expect all types need to be processed the same way using the workload I mentioned earlier. Am I missing something?

[the remark this answers is not in the document]
Comment (kagamiori) [reply velox/expression/CastExpr.cpp:556]: I'm not sure why either because those are existing code. I guess it's because peeling off encoding improves the performance of processing complex-typed data more significantly, so it was worth doing.

[the remark this answers is not in the document]
Comment (mbasmanova) [reply velox/expression/CastExpr.cpp:556]: Another question is whether expression evaluation should peel off encodings and exclude nulls before calling evalSpecialForm. If it did that, then special forms like CAST would not need to duplicate that logic.

CC: @bikramSingh91

[the remark this answers is not in the document]
Comment (bikramSingh91) [reply velox/expression/CastExpr.cpp:556]: To answer this question, I first tried to look as to why we need special form expr in the first place. Its because they don’t follow the regular flow of expr eval as defined in expr.cc and we can arbitrarily handle inputs as they wish.
This works perfectly for others like conjuncts or coalesce but cast seems to be the only one which has a flow very similar to the regular flow but with some added steps like extracting internal vectors from complex types. I believe adding a mechanism to allow special forms to reuse the flow might end up with a solution over-fitted to cast only, so might be an overkill unless we have more use cases like this. However, I feel there is definitely value in modularizing functionality like peeling and null handling in evalEncodings and evalWithNulls respectively that we can re-use here. The functionality in evalEncodings is also repeated somewhat in applyFunctionWithPeeling and hence can also be reused there. Moreover, like you suggested earlier, we can test those modular functions separately to make sure their functionality is bullet proof

File: velox/expression/Expr.cpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (kagamiori) [reply velox/expression/Expr.cpp]: Inlined these functions in the callers according to the offline discussion.
